Sort menus in the backend menu alphabetically

Hackwar opened the pull request:
This PR sorts the menutypes in the backend in alphabetical order and not in order how they were first created some ancient time ago. :wink:
### How to test
1. Go into the backend, open the "Menus" menu and see that it is not ordered alphabetically.
2. Apply change.
3. See that the menus are now ordered alphabetically.

Hackwar committed:
Sorting menutypes in the backend menu

zero-24 commented on administrator/modules/mod_menu/tmpl/default_enabled.php:
```diff
@@ -139,7 +139,8 @@
 	// Menu Types
-	foreach (ModMenuHelper::getMenus() as $menuType)
+	$menuTypes = JArrayHelper::sortObjects(ModMenuHelper::getMenus(), 'title');
+	foreach ($menuTypes as $menuType)
```
@Hackwar can you add a line bevor the `foreach`?

Hackwar replied:
Done.

Hackwar committed:
Codestyle

richard67 commented:
@test Success.
![screen shot 2015-01-12 at 14 14 00](http://issues.joomla.org/uploads/1/2ed61e703e13af94ca7b12ad206b98f2.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5683">issues.joomla.org/joomla-cms/5683</a>.</sub>

brianteeman commented:
I get the following error

`Strict Standards: Only variables should be passed by reference in /Applications/MAMP/htdocs/joomla-cms-staging/administrator/modules/mod_menu/tmpl/default_enabled.php on line 142`
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5683">issues.joomla.org/joomla-cms/5683</a>.</sub>

richard67 commented:
Ooops: I have the same, haven't seen it when testing before (blush).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5683">issues.joomla.org/joomla-cms/5683</a>.</sub>

Hackwar committed:
Fixing notice

Hackwar commented:
Sorry, fixed that one. I even looked at the source of JArrayHelper::sortObjects() and saw the reference, but missed it...

richard67 commented:
@test Success. Works now without PHP warnings after correction.
What would be perfect of course wold be to have an ordering column for the menus, as we have for the menu items, and show them in this order in the menu (which can be modified by drag and drop in the menu manager).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5683">issues.joomla.org/joomla-cms/5683</a>.</sub>

brianteeman commented:
@test the Strict standards issue has now been fixed. 

I now see that this change isnt applied in Hathor. Not sure if that uses an override ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5683">issues.joomla.org/joomla-cms/5683</a>.</sub>

richard67 commented:
@brianteeman  Yes, I just see Hathor has an override mod_menu/default_enabled.php.
@test I've forgotten Hathor (blush).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5683">issues.joomla.org/joomla-cms/5683</a>.</sub>

infograf768 commented:
Not sure I like this one. In my multilingual test site, the menus which were created one language after the other are now disordered.

Also, the sorting done in this PR is different in mod_menu vs the dropdown in "Menu Manager: Menu items"
In mod_menu all menus starting with a Capital letter are grouped first, not in the dropdown where it it is not case sensitive, i.e. 

<b>mod_menu</b>
Mainmenu
Top menu
[...]
my menu
toys menu

<b>Dropdown</b>
Mainmenu
my menu
[...]
Top menu
toys menu

If it was decided to alpha sort the menus, the second solution is much better imho

can be used instead, to solve at least the case sensitive issue (IF we sort alpha)
`$menuTypes = JArrayHelper::sortObjects($menuTypes, 'title', $direction = 1, $caseSensitive = false, $locale = true);`
or, shorter
`$menuTypes = JArrayHelper::sortObjects($menuTypes, 'title', 1, false, true);`

we would get

![screen shot 2015-01-13 at 09 04 27](https://cloud.githubusercontent.com/assets/869724/5717568/44d54ecc-9b03-11e4-81ac-e728659e6824.png)

instead of

![screen shot 2015-01-13 at 09 14 18](https://cloud.githubusercontent.com/assets/869724/5717648/a59965d0-9b04-11e4-8d62-855c28738a47.png)


changing locale to false has apparently no effect here.

Hackwar committed:
Adding Hathor support and making sorting case-insensitive

Hackwar committed:
Removing locale parameter again

Hackwar commented:
I've changed the sorting to be case-insensitive and added the change to Hathor, too. I also first thought I enabled the local-aware sorting, but then noticed that that feature is actually implemented pretty stupidly... I would have to hand in the locale by which to sort and not simply true or false and quite frankly, the amount of code that would have to be written to do this properly is simply not worth the gain we maybe would get. As far as I can see, it would only make sure that letters like o and ö would be grouped next to each other in German for example. However, I think that it already orders via UTF8 and I guess that that takes care of this already anyway... Mind you, that is not based on hard facts, but a hunch. :wink: 

In any case, I guess this is now ready for final testing/merging, depending on if this is actually a wanted feature. However, if we don't do this, we should remove the ordering in the menu manager filter dropdown to keep this consistent.

infograf768 commented:
I agree it should be consistent, thus why I proposed the code.

brianteeman commented:
@test all works well for me - thanks for this @Hackwar 

> However, if we don't do this, we should remove the ordering in the menu manager filter dropdown to keep this consistent.

I dont see this filter?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5683">issues.joomla.org/joomla-cms/5683</a>.</sub>

Hackwar commented:
if you go into the list view of a menu in the menu manager, you can switch between the menus that are displayed. That is the dropdown that I'm talking about.

brianteeman commented:
Ah gotcha
On 13 Jan 2015 10:18, "Hannes Papenberg" <notifications@github.com> wrote:

> if you go into the list view of a menu in the menu manager, you can switch
> between the menus that are displayed. That is the dropdown that I'm talking
> about.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/5683#issuecomment-69723145>.
>

infograf768 commented:
Fine here. Merging. Thanks.
